common/flatpak-dir.c

```diff
@@ -7982,6 +8068,105 @@ flatpak_dir_find_installed_ref (FlatpakDir   *self,
   return NULL;
 }
 
+/* Given a list of refs in local_refspecs, remove any refs that have already
+ * been deployed and return a new GPtrArray containing only the undeployed
+ * refs. This is used by flatpak_dir_cleanup_undeployed_refs to determine
+ * which undeployed refs need to be removed from the local repository.
+ *
+ * Returns: (transfer-full): A #GPtrArray
+ */
+static GPtrArray *
+filter_out_deployed_refs (FlatpakDir *self,
+                          GPtrArray  *local_refspecs,
+                          GError    **error)
+{
+  g_autoptr(GPtrArray) undeployed_refs = g_ptr_array_new_full (local_refspecs->len, g_free);
+  g_autoptr(GVariant) deploy_data = NULL;
+  gsize i = 0;
+
+  for (; i < local_refspecs->len; ++i)
+    {
+      const gchar *refspec = g_ptr_array_index (local_refspecs, i);
+      g_autofree gchar *remote = NULL;
+      g_autofree gchar *ref = NULL;
+
+      if (!ostree_parse_refspec (refspec, &remote, &ref, error))
+        return FALSE;
+
+      g_autoptr(GVariant) deploy_data = flatpak_dir_get_deploy_data (self,
+                                                                     ref,
+                                                                     NULL,
+                                                                     NULL);
+
+      if (!deploy_data)
+        g_ptr_array_add (undeployed_refs, g_strdup (refspec));
+    }
+
+  return g_steal_pointer (&undeployed_refs);
+}
+
+/**
+ * flatpak_dir_cleanup_undeployed_refs:
+ *
+ * Find all flatpak refs in the local repository which have not been deloyed
+ * in the dir and remove them from the repository. You might want to call this
+ * function if you pulled refs into the dir but then decided that you did
+ * not want to deploy them for some reason. Note that this does not prune
+ * objects bound to the cleaned up refs from the underlying OSTree repository,
+ * you should consider using flatpak_dir_prune to do that.
+ *
+ * @self: a #FlatpakDir
+ * @cancellable: (allow-none): a #GCancellable
+ * @error: (allow-none): a #GError
```

Comment:
See previous comments about `(allow-none)` and about putting argument documentation at the top of the gtk-doc comment.